chapters/VK_EXT_present_timing/PresentTimeInfo.txt

```diff
@@ -0,0 +1,184 @@
+    image at any time.
+  * pname:idealPresentTime provides an indication to the presentation
+    engine of what the application desires its IPD to become.
+    An application can: set pname:idealPresentTime to the same value
+    as pname:targetPresentTime, indicating that the application
+    desires its IPD to remain steady.
+    A value earlier than
+    pname:targetPresentTime indicates that the application desires to
+    shorten its IPD.
+    A value later than pname:targetPresentTime
+    indicates that the application desires to lengthen its IPD.
+    The presentation engine will provide feedback on this value in the
+    data returned by flink:vkGetPastPresentationTimingEXT.
+  * pname:presentSlop is the period of time in nanoseconds before
+    pname:targetPresentTime that the presentation engine may: display
+    the image.
+
+If pname:targetPresentTime is non-zero, the presentation engine should:
+not display the image to the user at a time earlier than
+(pname:targetPresentTime - pname:presentSlop).
+
+Setting pname:targetPresentTime to zero is useful when the application
+desires to provide a non-zero pname:presentID in order to get back timing
+information from flink:vkGetPastPresentationTimingEXT, but does not wish
+to set a pname:targetPresentTime.
+
+Providing a pname:presentSlop enables the application to indicate to the
+presentation engine to display the image within this period if a vertical
+blanking period occurs on FRR displays.
+
+[NOTE]
+.Note
+====
+The pname:presentSlop is used to avoid unintentionally missing a vertical
+blanking period on FRR displays due to rounding errors or drift between
+clocks.
+A suggested value for pname:presentSlop is half the _Refresh Rate_.
+====
+
+include::{generated}/validity/structs/VkAbsolutePresentTimeEXT.txt[]
+
+--
+
+[open,refpage='VkRelativePresentTimeEXT',desc='Specifying the minimum duration an image should be presented',type='structs']
+--
+
+The sname:VkRelativePresentTimeEXT structure is defined as:
+
+include::{generated}/api/structs/VkRelativePresentTimeEXT.txt[]
+
+  * pname:presentID is an application-provided identification value.
+    If non-zero, flink:vkGetPastPresentationTimingEXT will return a
+    slink:VkPastPresentationTimingEXT struct with the timing information
+    associated with the present to the associated swapchain for this
+    call to flink:vkQueuePresentKHR, including this value being in the
+    pname:presentID member of the slink:VkPastPresentationTimingEXT
+    struct.
+    In order to be useful to the application, it should: be unique within
+    some period of time that is meaningful to the application.
+  * pname:minPresentDuration, if non-zero, specifies the minimum duration
+    in nanoseconds the application wants the image to be displayed to the
+    user.
+    A value of zero specifies that the presentation engine may: display the
+    image for any duration.
+  * pname:idealPresentDuration provides an indication to the presentation
```

Comment:
The exact semantics of idealPresentDuration seem a bit vague to me. I don't quite understand how it works.

There seem to be three distinct cases for it:

target == ideal:

Can optimalPresent still be smaller than target in case where we could go from 33.3ms to 16.67ms duration? If so, why use ideal at all?

target < ideal:

Does this mean anything? Will the presentation engine try to aim for present at ideal instead, rather than target? If so, can't the app just use target = ideal?

target > ideal:

Same question as the first case with equality. Are optimalPresent values returned somehow directly equal to ideal here? I.e., will driver look at reported present time, and snap it to idealTime if it's close enough? If so, what is close enough for snapping to happen?

Reply:
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/45/diffs#f852003dd5091841baf96e797d61195b312f3395_0_22 seems more precisely specified, if it's intended that it should work like the Wayland RFC.